src/main/scala/mesosphere/marathon/core/storage/impl/zk/ZkPersistenceStore.scala

```diff
@@ -0,0 +1,147 @@
+package mesosphere.marathon.core.storage.impl.zk
+
+import java.time.OffsetDateTime
+
+import akka.actor.Scheduler
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.{ Done, NotUsed }
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.StoreCommandFailedException
+import mesosphere.marathon.core.storage.CategorizedKey
+import mesosphere.marathon.core.storage.impl.BasePersistenceStore
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.util.{ Retry, toRichFuture }
+import mesosphere.util.state.zk.{ Children, GetData, RichCuratorFramework }
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.NoNodeException
+import org.apache.zookeeper.data.Stat
+
+import scala.async.Async.{ async, await }
+import scala.collection.immutable.Seq
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success }
+
+class ZkPersistenceStore(val client: RichCuratorFramework)(
+    implicit
+    mat: Materializer,
+    ctx: ExecutionContext,
+    scheduler: Scheduler,
+    val metrics: Metrics
+) extends BasePersistenceStore[ZkId, String, ZkSerialized]() with StrictLogging {
+
+  private val retryOn: Retry.RetryOnFn = {
+    case _: KeeperException.ConnectionLossException => true
+    case _: KeeperException => false
+    case NonFatal(_) => true
+  }
+
+  private def retry[T](name: String)(f: => Future[T]) = Retry(name, retryOn = retryOn)(f)
+
+  override protected def rawIds(category: String): Source[ZkId, NotUsed] = {
+    val childrenFuture = retry(s"ZkPersistenceStore::ids($category)") {
+      async {
+        val buckets = await(client.children(s"/$category").recover {
+          case _: NoNodeException => Children(category, new Stat(), Nil)
+        }).children
+        val children = await(Future.sequence(buckets.map(b => client.children(s"/$category/$b").map(_.children))))
```

Comment:
I'm always a bit frightened by using Future.sequence while hitting zk (we had some P0 exactly for that). We definitively need a specific ExecutionContext for zk with a limited thread pool (Something like n=8 ?).